lib/linkerd2-error/src/recover.rs

```diff
@@ -0,0 +1,63 @@
+use super::{Error, Never};
+use futures::{stream, Stream};
+
+/// An error recovery strategy.
+pub trait Recover<E: Into<Error> = Error> {
+    type Error: Into<Error>;
+    type Backoff: Stream<Item = (), Error = Self::Error>;
+
+    /// Given an E-typed error, determine if the error is recoverable.
+    ///
+    /// If it is, a backoff stream is returned. When the backoff becomes ready,
+    /// it signals that the caller should retry its operation. If the backoff is
+    /// polled agian, it is assumed that the operation failed and a new (possibly
+    /// longer) backoff is initated.
+    ///
+    /// If the error is not recoverable, it is returned immediately.
+    fn recover(&self, err: E) -> Result<Self::Backoff, E>;
+}
+
+#[derive(Copy, Clone, Debug, Default)]
+pub struct Immediately(());
+
+// === impl Recover ===
+
+impl<E, B, F> Recover<E> for F
```

Comment:
Is this impl here only for development use? I'm good with unused code and it doesn't seem to be used right now, but could possibly be helpful when writing a recover impl for a `Fn` or something?

Reply:
Since `Recover` is really just a way to have a name for a `FnMut(E) -> Result<Backoff, E>`, it's convenient to provide an implementation. In fact, we can/should probably use this to get rid of the `always` stuff

Reply:
Oh cool nice refactor to get rid of `always`. Looks good